### PR TITLE
Modest performance, address #12647

### LINKED ISSUE #12647: TypeError with pd.read_sas() (pandas 0.18.0)
#### Code Sample, a copy-pastable example if possible
df = pd.read_sas('infilename.sas7bdat')
#### Expected Output
to read a sas7bdat file into a pandas data frame.
#### output of ``pd.show_versions()``
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.11.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 69 Stepping 1, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None

pandas: 0.18.0
nose: 1.3.7
pip: 8.1.0
setuptools: 20.2.2
Cython: 0.23.4
numpy: 1.10.4
scipy: 0.17.0
statsmodels: 0.6.1
xarray: None
IPython: 4.0.3
sphinx: 1.3.5
patsy: 0.4.0
dateutil: 2.5.0
pytz: 2016.1
blosc: None
bottleneck: 1.0.0
tables: 3.2.2
numexpr: 2.4.6
matplotlib: 1.5.1
openpyxl: 2.3.2
xlrd: 0.9.4
xlwt: 1.0.0
xlsxwriter: 0.8.4
lxml: 3.5.0
bs4: 4.4.1
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.11
pymysql: None
psycopg2: None
jinja2: 2.8
boto: 2.39.0
---------------------------------------------------------------------------
### Error seen

TypeError                                 Traceback (most recent call last)
<ipython-input-9-3046a8f15981> in <module>()
----> 1 pd.read_sas('mydatainfo.sas7bdat')

C:\Anaconda\lib\site-packages\pandas\io\sas\sasreader.py in read_sas(filepath_or_buffer, format, index, encoding, chunksize, iterator)
     52         reader = SAS7BDATReader(filepath_or_buffer, index=index,
     53                                 encoding=encoding,
---> 54                                 chunksize=chunksize)
     55     else:
     56         raise ValueError('unknown SAS format')

C:\Anaconda\lib\site-packages\pandas\io\sas\sas7bdat.py in __init__(self, path_or_buf, index, convert_dates, blank_missing, chunksize, encoding)
    234             self._path_or_buf = open(self._path_or_buf, 'rb')
    235 
--> 236         self._get_properties()
    237         self._parse_metadata()
    238 

C:\Anaconda\lib\site-packages\pandas\io\sas\sas7bdat.py in _get_properties(self)
    333             self.os_name = buf.rstrip(b'\x00 ').decode()
    334         else:
--> 335             buf = self._path_or_buf.read(_os_maker_offset, _os_maker_length)
    336             self.os_name = buf.rstrip(b'\x00 ').decode()
    337 

TypeError: read() takes at most 1 argument (2 given)

---------------------------------------------------------------------------
## Tracking down the error:

It looks like line 335 should either be set up to read a single length from the _path_or_buf bytestream something like:
``` buf = self._path_or_buf.read(_os_maker_length) ``` 
or replace it with something like this
``` buf = self._read_bytes(_os_maker_offset, _os_maker_length) ```

---------------------------------------------------------------------------
** note i tried this under python3.5 and got the same error.

### LINKED ISSUE #12654: Read SAS  (sas7bdat)
Very excited to have this new feature in Pandas. I have a few comments to share:

1. `pd.read_sas()` doesn't read SAS date variable correctly (this is noted in the doc). Dates are read as `numpy.float64`. Note in SAS, dates are recorded as numbers relative to 1960-1-1. It would be helpful to allow some sort of arguments to parse the date variable correctly. 

2. Moreover, SAS has some special missing variables such as `.B` or `.R`. I wonder how are these cases treated? 

3. Not nearly as fast as `read_csv()`. To read a 700MB SAS data. The time is
```
CPU times: user 1min 47s, sys: 955 ms, total: 1min 48s
Wall time: 1min 48s
```
The time for the same CSV file (I covered the same file to CSV using SAS) is 
```
CPU times: user 3.93 s, sys: 343 ms, total: 4.28 s
Wall time: 4.29 s
```

### LINKED ISSUE #12659: BUG: Read SAS fails
cc: @kshedden

#### Code Sample, a copy-pastable example if possible

https://github.com/pydata/pandas/blob/8857008178eb90b59ca67874507ef860b3faf88f/pandas/io/tests/sas/data/test17.sas7bdat

Do you want a PR with this test case?

#### Expected Output

pandas reads the file without crashing :)
I tracked the problem to _process_columntext_subheader. It seems like there is an encoding issue...

You might need to apply  PR #12658 first to see this issue.

FWIW, here is an AWFUL HACK to make pandas read the file "almost" correctly (I get column names as bytes instead of str in this case).

```diff
@@ -524,17 +524,17 @@ class SAS7BDATReader(BaseIterator):
 
         offset += self._int_length
         text_block_size = self._read_int(offset, _text_block_size_length)
 
         buf = self._read_bytes(offset, text_block_size)
-        self.column_names_strings.append(
-            buf[0:text_block_size].rstrip(b"\x00 ").decode())
+        self.column_names_strings.append(buf)
         if len(self.column_names_strings) == 1:
             column_name = self.column_names_strings[0]
             compression_literal = ""
             for cl in _compression_literals:
+                cl = bytes(cl, 'ascii')
                 if cl in column_name:
                     compression_literal = cl
             self.compression = compression_literal
             offset -= self._int_length
```

#### output of ``pd.show_versions()``

commit: None
python: 3.5.1.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 69 Stepping 1, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None

pandas: 0.18.0
nose: 1.3.7
pip: 8.1.0
setuptools: 20.2.2
Cython: None
numpy: 1.10.4
scipy: None
statsmodels: None
xarray: None
IPython: 4.1.2
sphinx: 1.3.1
patsy: None
dateutil: 2.4.2
pytz: 2015.7
blosc: None
bottleneck: 1.0.0
tables: 3.2.2
numexpr: 2.4.6
matplotlib: 1.5.1
openpyxl: 2.3.2
xlrd: 0.9.4
xlwt: None
xlsxwriter: 0.8.4
lxml: 3.5.0
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.8
boto: None

### LINKED ISSUE #12809: Pandas read_sas error: 'ascii' codec can't decode byte 0xd8 in position 0: ordinal not in range(128)
Hello everybody,

I am using **Pandas 0.18** to open a `sas7bdat` dataset 

I simply use:

    df=pd.read_sas('P:/myfile.sas7bdat')

and I get the following error

        buf[0:text_block_size].rstrip(b"\x00 ").decode())
    
    UnicodeDecodeError: 'ascii' codec can't decode byte 0xd8 in position 0: ordinal not in range(128)

If I use

    import sys
    reload(sys)
    sys.setdefaultencoding("utf-8")

I get

    UnicodeDecodeError: 'utf8' codec can't decode byte 0xd8 in position 0: invalid continuation byte

Other `sas7bdat` files in my folder are handled just fine by Pandas. 

When I open the file in SAS I see that the column names are very long and span several lines, but otherwise the files look just fine.

There are not so many possible options in `read_sas`... what should I do? Is this a bug in `read_sas`? 

Many thanks!

### PR DESCRIPTION
**Pull request by kshedden**
closes #12659
closes #12654
closes #12647
closes #12809 

Major performance improvements through use of Cython

### CODE REVIEW

**Comment by jreback** [comment]
@kshedden you prob know this, but all of the perf issues have to do with going back-forth between cython/python. the ideal would be to put the entire loop in cython, rather than in python, then a call to parse a single line in cython.

**Comment by kshedden** [comment]
Actually I didn't know that there was that much function call overhead for
cython.  I am just passing scalar value and memory slices (plus the parser
object but we could refactor that out if that is the problem).

**Comment by jreback** [comment]
yeah because everything has to be checked every time. Its not what you are passing, but the function call itself. You should be able move the entire loop and will get pretty good speedups.

**Comment by kshedden** [comment]
This should resolve #12659, #12654, #12647.  Also adds test coverage with two new files.

The tests pass on most setups but there is one coredump on Travis.  I'm not sure it that is related to this PR.

There are also some performance enhancements here in the 2x-4x range.

@gdementen @jreback @ywhuofu @benjello @raderaj

File: pandas/io/sas/saslib.pyx [hunk 1]
```diff
@@ -191,43 +193,132 @@ def _rdc_decompress(int result_length, np.ndarray[uint8_t, ndim=1] inbuff):
...
+def _do_read(parser, int nrows):
+    cdef int i
+
+    for i in range(nrows):
+        done = _readline(parser)
+        if done:
+            break
+
+
+def _readline(parser):
```
Comment (jreback) [inline pandas/io/sas/saslib.pyx]: make this a cdef

File: pandas/io/sas/saslib.pyx [hunk 2]
```diff
@@ -191,43 +193,132 @@ def _rdc_decompress(int result_length, np.ndarray[uint8_t, ndim=1] inbuff):
...
 
-def process_byte_array_with_data(parser, int offset, int length, np.ndarray[uint8_t, ndim=2] byte_chunk,
-                                 np.ndarray[dtype=object, ndim=2] string_chunk):
+
+def _do_read(parser, int nrows):
+    cdef int i
+
+    for i in range(nrows):
+        done = _readline(parser)
+        if done:
```
Comment (jreback) [inline pandas/io/sas/saslib.pyx]: type everything (done should be a bint)

File: pandas/io/sas/saslib.pyx (hunk 1 above)
Comment (jreback) [inline pandas/io/sas/saslib.pyx]: cdef bint _readline

File: pandas/io/sas/saslib.pyx [hunk 3]
```diff
@@ -191,43 +193,132 @@ def _rdc_decompress(int result_length, np.ndarray[uint8_t, ndim=1] inbuff):
 
     return np.asarray(outbuff).tostring()
 
-def process_byte_array_with_data(parser, int offset, int length, np.ndarray[uint8_t, ndim=2] byte_chunk,
-                                 np.ndarray[dtype=object, ndim=2] string_chunk):
+
+def _do_read(parser, int nrows):
```
Comment (jreback) [inline pandas/io/sas/saslib.pyx]: you don't have to, but its good to do: ``_do_read(object parser, int nrows):``

File: pandas/io/sas/saslib.pyx [hunk 4]
```diff
@@ -191,43 +193,132 @@ def _rdc_decompress(int result_length, np.ndarray[uint8_t, ndim=1] inbuff):
...
+    for i in range(nrows):
+        done = _readline(parser)
+        if done:
+            break
+
+
+def _readline(parser):
+
+    cdef int offset
+    cdef int bit_offset
```
Comment (jreback) [inline pandas/io/sas/saslib.pyx]: you can declare things like:

```
cdef:
        int offset, bit_offset, ......
```

File: pandas/io/sas/saslib.pyx [hunk 5]
```diff
@@ -191,43 +193,132 @@ def _rdc_decompress(int result_length, np.ndarray[uint8_t, ndim=1] inbuff):
...
+    if parser._cached_page is None:
+        parser._path_or_buf.seek(parser.header_length)
+        done = parser._read_next_page()
+        if done:
+            return True
+
+    # Loop until a data row is read
+    while True:
+        if parser._current_page_type == const.page_meta_type:
+            flag = (parser._current_row_on_page_index >=
```
Comment (jreback) [inline pandas/io/sas/saslib.pyx]: in order to get some real perf increases. you are going to actually have to keep these variables as typed in cython.

File: pandas/io/sas/saslib.pyx [hunk 6]
```diff
@@ -191,43 +193,132 @@ def _rdc_decompress(int result_length, np.ndarray[uint8_t, ndim=1] inbuff):
...
+                    len(parser._current_page_data_subheader_pointers))
+            if flag:
+                done = parser._read_next_page()
```
Comment (jreback) [inline pandas/io/sas/saslib.pyx]: you need to move routines like this to cython as well.

File: pandas/io/sas/saslib.pyx [hunk 7]
```diff
@@ -191,43 +193,132 @@ def _rdc_decompress(int result_length, np.ndarray[uint8_t, ndim=1] inbuff):
...
+                if done:
+                    return True
+                parser._current_row_on_page_index = 0
+            return False
+        else:
+            raise ValueError("unknown page type: %s",
+                             parser._current_page_type)
+
+
+def process_byte_array_with_data(parser, int offset, int length, uint8_t[:, ::1] byte_chunk,
```
Comment (jreback) [inline pandas/io/sas/saslib.pyx]: needs to be a cdef and have a return type.

**Comment by kshedden** [comment]
@jreback I'm having trouble working with object dtype arrays in cython, would appreciate your help.

I have the following code, where `source` is a memory slice directly copied from the SAS file, and I need to slice it into smaller chunks which correspond to individual data values.  I am able to do something similar fine for float64 types (not shown below), but I can't get it to work for variable-length byte arrays.  These byte arrays will eventually become strings but I don't want to do the conversion here because we give the user the option to retain the raw bytes.  The error message I am getting is copied below the code.  

I have also tried typing `string_chunk` as `object[:, ::1]` but that didn't work either.

```
cdef void process_byte_array_with_data(object parser, int offset, int length,
                                       uint8_t[:, ::1] byte_chunk,
                                       np.ndarray string_chunk):
    # ...
    bvec = bytearray(source[start:start+lngt])
    string_chunk[js, parser._current_row_in_chunk_index] = bvec
```

```
BufferError: Object is not writable.
Exception ignored in: 'pandas.io.sas.saslib.process_byte_array_with_data'
Traceback (most recent call last):
  File "stringsource", line 616, in View.MemoryView.memoryview_cwrapper (pandas/io/sas/saslib.c:14511)
  File "stringsource", line 323, in View.MemoryView.memoryview.__cinit__ (pandas/io/sas/saslib.c:10880)
```

**Comment by jreback** [comment]
docs are [here](http://docs.cython.org/src/tutorial/strings.html)

I think you want something like:

```
def process_byte_data(unsigned char[:] data):
    length = data.shape[0]
    first_byte = data[0]
    slice_view = data[1:-1]
```

so your ``data`` would be your source, its just a bunch of bytes.
you can directly slice them. Then you can cast them to what you need (you need to allocate memory and such). This is very much like working directly in C.

note that you are *always* working with bytes, and NOT unicode/str. you can decode things later.

HTH, and lmk.

**Comment by kshedden** [comment]
@jreback I think this is ready now, known bugs and performance issues should be resolved

**Comment by jreback** [comment]
can u post perf comparison (also haven't looked but do we have asv for this?)

**Comment by kshedden** [comment]
We don't have an asv, I tried setting one up but failed (I don't have a continuum distro and couldn't get it to work with virtualenv).

A rough timing on a 100K file: the released version (v0.18) takes around 11 seconds, this version takes around 5 seconds.  However csv reading is about 50x faster, so still a ways to go.

**Comment by jreback** [comment]
how this coming?

**Comment by kshedden** [comment]
I think it is ready. to go, the known bugs are fixed and the performance is at least somewhat better.

File: doc/source/whatsnew/v0.18.1.txt [hunk 8]
```diff
@@ -191,7 +191,7 @@ Deprecations
 Performance Improvements
 ~~~~~~~~~~~~~~~~~~~~~~~~
 
-
+- Improved speed of SAS reader (PR 12656)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: write the issue number like others (it doesn't matter that its a PR number)

File: doc/source/whatsnew/v0.18.1.txt [hunk 9]
```diff
@@ -215,9 +215,9 @@ Bug Fixes
 - Bug in ``GroupBy.filter`` when ``dropna=False`` and no groups fulfilled the criteria (:issue:`12768`)
 - Bug in ``__name__`` of ``.cum*`` functions (:issue:`12021`)
 
-
 - Bug in ``.drop()`` with a non-unique ``MultiIndex``. (:issue:`12701`)
 - Bug in ``.concat`` of datetime tz-aware and naive DataFrames (:issue:`12467`)
+- Various encoding and header processing issues in SAS reader (:issue:`12659`, :issue:`12654`, :issue:`12647`, :issue:`12809`, closed by PR 12656)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: you don't need the 'closed by ....'

File: pandas/io/sas/sas7bdat.py [hunk 10]
```diff
@@ -202,18 +53,27 @@ class SAS7BDATReader(BaseIterator):
         Return SAS7BDATReader object for iterations, returns chunks
         with given number of lines.
     encoding : string, defaults to None
-        String encoding.  If None, text variables are left as raw bytes.
+        String encoding.
+    convert_text : bool, defaults to True
```
Comment (jreback) [inline pandas/io/sas/sas7bdat.py]: add a versionadded directive

File: pandas/io/sas/sas7bdat.py [hunk 11]
```diff
@@ -202,18 +53,27 @@ class SAS7BDATReader(BaseIterator):
...
 
         self.index = index
         self.convert_dates = convert_dates
         self.blank_missing = blank_missing
         self.chunksize = chunksize
         self.encoding = encoding
+        self.convert_text = convert_text
+        self.convert_header_text = convert_header_text
 
+        self.default_encoding = "latin-1"
```
Comment (jreback) [inline pandas/io/sas/sas7bdat.py]: why is this not the default for the ``encoding`` argument?

**Comment by jreback** [comment]
a PEP checked faile

``git diff master | flake8 --diff``

**Comment by jreback** [comment]
I assume this includes the fixes from #12658 (or equivalent) as well?

**Comment by jreback** [comment]
FYI: http://www.principlesofeconometrics.com/sas.htm (getting encoding errors on the first one I tried, airlines...)

**Comment by kshedden** [comment]
@jreback the file you linked works for me in py3 not py2, will look into that

**Comment by jreback** [comment]
yeah master seemed to have a Unicode issue with one of the files - not sure why it didn't show up on Travis

**Comment by kshedden** [comment]
@jreback I think this is ready now

**Comment by jreback** [comment]
did u look at the refactor I posted?

**Comment by kshedden** [comment]
Do you mean this one?

https://github.com/pydata/pandas/pull/12656#issuecomment-207865144

**Comment by jreback** [comment]
https://github.com/kshedden/pandas/pull/1

**Comment by kshedden** [comment]
Thanks, runs clean locally, waiting on travis now.  The only change I made is that I left a few things in the decompressors int type due to overflow potential.

**Comment by jreback** [comment]
btw I only had a really small benchmark file
you prob have better

File: pandas/io/sas/saslib.pyx [hunk 12]
```diff
@@ -189,49 +180,205 @@ def _rdc_decompress(int result_length, np.ndarray[uint8_t, ndim=1] inbuff):
...
+        if self.cached_page == NULL:
+            self.parser._path_or_buf.seek(self.parser.header_length)
+            done = self.read_next_page()
+            if done:
+                return True
+
+        # Loop until a data row is read
+        while True:
+            if self.parser._current_page_type == const.page_meta_type:
+                flag = (self.current_row_on_page_index >=
```
Comment (jreback) [inline pandas/io/sas/saslib.pyx]: FYI there is a fair amount more that can be typed here but would involve some rewriting

**Comment by kshedden** [comment]
For the test file I have been using, I am now getting 3 seconds per 100K lines verus 11 seconds per 100K lines in the 0.18.0 version.

Also, I'm not aware of any files that fail under this version, although in some cases the encoding will need to be manually specified.

**Comment by jreback** [comment]
oh that sounds great!

oh so you are able to infer encodings in some instances? worth mentioning in docs / doc-string?

are there any other updates for doc-string / docs?

**Comment by jreback** [comment]
couple of issues on windows. going to fixup.

**Comment by jreback** [comment]
the airlines.sas7bdat, csv seem to be missing, can you push them up in another commit?

**Comment by kshedden** [comment]
done

**Comment by jreback** [comment]
ty. not sure why that didn't fail the tests, but oh well.

**Comment by jreback** [comment]
thanks @kshedden great effort!!!!

**Comment by jreback** [comment]
I may come back around with some more perf improvements in the cython code when I have some time.

**Pull request closed without merging**